File: Year2/TDA417-DataStructures_Algorithms/Labs/Lab1/work/python/binary_search.py
from abc import abstractmethod
from typing import Protocol, TypeVar, List

class ComparableProtocol(Protocol):
    """Protocol for annotating comparable types."""
    @abstractmethod
    def __lt__(self: 'Comparable', other: 'Comparable', /) -> bool: ...

Comparable = TypeVar('Comparable', bound=ComparableProtocol)


def containsIterative(array: List[Comparable], value: Comparable) -> bool:

    min = 0
    max = len(array)-1

    while min <= max:

        middle = (min + max) // 2

        if array[middle] < value:
            min = middle + 1
        elif array[middle] > value:
            max = middle - 1
        else:
            return True
    return False


def containsRecursive(array: List[Comparable], value: Comparable) -> bool:

    return containsRecursiveHelper(array, value, 0, len(array)-1)

def containsRecursiveHelper(array,value,min,max):

    if max == None:
        max = len(array)

    middle = (min+max) // 2

    if min > max:
        return False

    if array[middle] < value:
        min = middle+1
        return containsRecursiveHelper(array,value,min,max)
    elif array[middle] > value:
        max = middle-1
        return containsRecursiveHelper(array,value,min,max)
    else:
        return True


def firstIndexOf(array: List[Comparable], index: Comparable) -> int:
    min = 0 
    max = len(array)-1
    result = -1

    while min <= max:
        middle = (min+max) // 2

        if array[middle] == index:
            result = middle
            max = middle - 1
        elif array[middle] > index:
            max = middle - 1
        else:
            min = middle + 1
    return result
        

def main():
    integerTestArray = [1, 3, 5, 7, 9]

    assert containsIterative(integerTestArray, 4) == False
    assert containsIterative(integerTestArray, 7) == True
    assert containsRecursive(integerTestArray, 0) == False
    assert containsRecursive(integerTestArray, 9) == True

    stringTestArray = ["cat", "cat", "cat", "dog", "turtle", "turtle"]

    assert firstIndexOf(stringTestArray, "cat") == 0
    assert firstIndexOf(stringTestArray, "dog") == 3
    assert firstIndexOf(stringTestArray, "turtle") == 4
    assert firstIndexOf(stringTestArray, "zebra") == -1


if __name__ == '__main__':
    main()

File: Year2/TDA417-DataStructures_Algorithms/Labs/Lab1/work/java/BinarySearch.java

/**
 * Different implementations of binary search.
 * 
 * Common description of the below functions:
 * - Precondition: `array` is sorted according to the natural order.
 * - Precondition: all arguments are non-null (no need to check).
 * - Required complexity: O(log(n)) comparisons where n is the length of `array`.
*/
public class BinarySearch {

    /**
     * Check if the array contains the given value.
     * Iterative solution.
     * 
     * @param array an array sorted according to the natural order of its elements.
     * @param value the value to search for.
     * @return true if `value` is in `array`.
     */
    public static<V extends Comparable<? super V>> boolean containsIterative(V[] array, V value) {
        //---------- TASK 1: Iterative version of binary search -------------------//
        // Hint: you probably need some auxiliary variables telling which part 
        // of the array you're looking at.
        throw new UnsupportedOperationException("TODO");
        //---------- END TASK 1 ---------------------------------------------------//
    }

    /**
     * Check if the array contains the given value.
     * Recursive solution.
     * 
     * @param array an array sorted according to the natural order of its elements.
     * @param value the value to search for.
     * @return true if `value` is in `array`.
     */
    public static<V extends Comparable<? super V>> boolean containsRecursive(V[] array, V value) {
        //---------- TASK 2: Recursive version of binary search -------------------//
        // Hint: you need a recursive helper function with some extra
        // arguments telling which part of the array you're looking at.
        throw new UnsupportedOperationException("TODO");
        //---------- END TASK 2 ---------------------------------------------------//
    }

    /**
     * Return the *first* position in the array whose element matches the given value.
     * 
     * @param array an array sorted according to the natural order of its elements.
     * @param value the value to search for.
     * @return the first position where `value` occurs in `array`, or -1 if it doesn't occur.
     */
    public static<V extends Comparable<? super V>> int firstIndexOf(V[] array, V value) {
        //---------- TASK 3: Binary search returning the first index --------------//
        // It's up to you if you want to use an iterative or recursive version.
        throw new UnsupportedOperationException("TODO");
        //---------- END TASK 3 ---------------------------------------------------//
    }

    // Put your own tests here.

    public static void main(String[] args) {
        Integer[] integer_test_array = {1, 3, 5, 7, 9};

        assert containsIterative(integer_test_array, 4) == false;
        assert containsIterative(integer_test_array, 7) == true;

        assert containsRecursive(integer_test_array, 0) == false;
        assert containsRecursive(integer_test_array, 9) == true;

        String[] string_test_array = {"cat", "cat", "cat", "dog", "turtle", "turtle"};

        assert firstIndexOf(string_test_array, "cat") == 0;
        assert firstIndexOf(string_test_array, "dog") == 3;
        assert firstIndexOf(string_test_array, "turtle") == 4;
        assert firstIndexOf(string_test_array, "zebra") == -1;
    }

}

